x/denommetadata/ibc_middleware.go

```diff
@@ -97,3 +92,20 @@ func (im IBCMiddleware) OnRecvPacket(
 
 	return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
 }
+
+func denomMetadataFromMemo(memo, ibcDenom string) (*banktypes.Metadata, error) {
+	transferInject := types.ParsePacketMetadata(memo)
+	if transferInject == nil || transferInject.DenomMetadata == nil {
+		return nil, nil
+	}
+	dm := transferInject.DenomMetadata
+
+	if err := dm.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid denom metadata: %w", err)
+	}
+
+	dm.Base = ibcDenom
+	dm.DenomUnits[0].Denom = dm.Base
```

Comment:
can you move it out of denomMetadataFromMemo to the callsite
I wouldn't expect this func to mutate anything